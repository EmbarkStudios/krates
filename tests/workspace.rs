mod util;

use util::build;

#[test]
fn includes() {
    let mut kb = krates::Builder::new();

    kb.include_workspace_crates([
        "/home/jake/code/krates/tests/ws2/b",
        "/home/jake/code/krates/tests/ws2/c/Cargo.toml",
    ]);

    let grafs = build("all-features2.json", kb).unwrap();
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn root() {
    // The ws2 workspace has a top level crate that is also a virtual manifest,
    // so it will have a resolution root, which will be used instead of the
    // list of workspace members, and since it doesn't depend on any of the
    // others in the workspace, it will be a graph of one
    let kb = krates::Builder::new();

    let grafs = build("all-features2.json", kb).unwrap();

    assert_eq!(grafs.actual.len(), 1);
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn workspace_with_root() {
    let mut kb = krates::Builder::new();
    // Setting the workspace true flag will mean to include all of the workspace
    // members, regardless of whether the resolution root is set or not
    kb.workspace(true);

    let grafs = build("all-features2.json", kb).unwrap();
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn workspace_with_root_exclude() {
    let mut kb = krates::Builder::new();
    kb.workspace(true);
    kb.exclude(std::iter::once("c".parse::<krates::PkgSpec>().unwrap()));

    let grafs = build("all-features2.json", kb).unwrap();
    insta::assert_snapshot!(grafs.dotgraph());
}
