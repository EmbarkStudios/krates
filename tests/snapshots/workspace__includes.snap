---
source: tests/workspace.rs
expression: grafs.dotgraph()
---
digraph {
    0 [ label = "crate b 0.1.0 (path+file:///home/jake/code/krates/tests/ws2/b)" ]
    1 [ label = "crate bindgen 0.59.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    2 [ label = "crate bitflags 1.3.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    3 [ label = "crate bumpalo 3.11.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    4 [ label = "crate c 0.1.0 (path+file:///home/jake/code/krates/tests/ws2/c)" ]
    5 [ label = "crate cc 1.0.73 (git+https://github.com/alexcrichton/cc-rs#53fb72c87e5769a299f1886ead831901b9c775d6)" ]
    6 [ label = "crate cc 1.0.73 (registry+https://github.com/rust-lang/crates.io-index)" ]
    7 [ label = "crate cexpr 0.6.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    8 [ label = "crate cfg-if 0.1.10 (registry+https://github.com/rust-lang/crates.io-index)" ]
    9 [ label = "crate cfg-if 1.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    10 [ label = "crate clang-sys 1.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    11 [ label = "crate coreaudio-rs 0.9.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    12 [ label = "crate coreaudio-sys 0.2.10 (registry+https://github.com/rust-lang/crates.io-index)" ]
    13 [ label = "crate difference 2.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    14 [ label = "crate glob 0.3.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    15 [ label = "crate js-sys 0.3.60 (registry+https://github.com/rust-lang/crates.io-index)" ]
    16 [ label = "crate lazy_static 1.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    17 [ label = "crate lazycell 1.3.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    18 [ label = "crate leftpad 0.2.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    19 [ label = "crate libc 0.2.134 (registry+https://github.com/rust-lang/crates.io-index)" ]
    20 [ label = "crate libloading 0.7.3 (registry+https://github.com/rust-lang/crates.io-index)" ]
    21 [ label = "crate log 0.4.17 (registry+https://github.com/rust-lang/crates.io-index)" ]
    22 [ label = "crate memchr 2.5.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    23 [ label = "crate minimal-lexical 0.2.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    24 [ label = "crate nix 0.16.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    25 [ label = "crate nom 7.1.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    26 [ label = "crate once_cell 1.15.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    27 [ label = "crate peeking_take_while 0.1.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    28 [ label = "crate proc-macro2 1.0.46 (registry+https://github.com/rust-lang/crates.io-index)" ]
    29 [ label = "crate quote 1.0.21 (registry+https://github.com/rust-lang/crates.io-index)" ]
    30 [ label = "crate regex 1.6.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    31 [ label = "crate regex-syntax 0.6.27 (registry+https://github.com/rust-lang/crates.io-index)" ]
    32 [ label = "crate ring 0.16.20 (registry+https://github.com/rust-lang/crates.io-index)" ]
    33 [ label = "crate rustc-hash 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    34 [ label = "crate shlex 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    35 [ label = "crate spin 0.5.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    36 [ label = "crate syn 1.0.101 (registry+https://github.com/rust-lang/crates.io-index)" ]
    37 [ label = "crate top 0.1.0 (path+file:///home/jake/code/krates/tests/ws2)" ]
    38 [ label = "crate unicode-ident 1.0.4 (registry+https://github.com/rust-lang/crates.io-index)" ]
    39 [ label = "crate untrusted 0.7.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    40 [ label = "crate void 1.0.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    41 [ label = "crate wasm-bindgen 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    42 [ label = "crate wasm-bindgen-backend 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    43 [ label = "crate wasm-bindgen-futures 0.4.33 (registry+https://github.com/rust-lang/crates.io-index)" ]
    44 [ label = "crate wasm-bindgen-macro 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    45 [ label = "crate wasm-bindgen-macro-support 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    46 [ label = "crate wasm-bindgen-shared 0.2.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    47 [ label = "crate web-sys 0.3.60 (registry+https://github.com/rust-lang/crates.io-index)" ]
    48 [ label = "crate winapi 0.2.8 (registry+https://github.com/rust-lang/crates.io-index)" ]
    49 [ label = "crate winapi 0.3.9 (registry+https://github.com/rust-lang/crates.io-index)" ]
    50 [ label = "crate winapi-i686-pc-windows-gnu 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    51 [ label = "crate winapi-x86_64-pc-windows-gnu 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    52 [ label = "feature default" ]
    53 [ label = "feature default" ]
    54 [ label = "feature default" ]
    55 [ label = "feature clang_6_0" ]
    56 [ label = "feature runtime" ]
    57 [ label = "feature std" ]
    58 [ label = "feature unicode" ]
    59 [ label = "feature default" ]
    60 [ label = "feature default" ]
    61 [ label = "feature default" ]
    62 [ label = "feature default" ]
    63 [ label = "feature Crypto" ]
    64 [ label = "feature Window" ]
    65 [ label = "feature std" ]
    66 [ label = "feature audio_toolbox" ]
    67 [ label = "feature audio_unit" ]
    68 [ label = "feature core_audio" ]
    69 [ label = "feature core_midi" ]
    70 [ label = "feature open_al" ]
    71 [ label = "feature runtime" ]
    72 [ label = "feature default" ]
    73 [ label = "feature errhandlingapi" ]
    74 [ label = "feature libloaderapi" ]
    75 [ label = "feature default" ]
    76 [ label = "feature extra_traits" ]
    77 [ label = "feature default" ]
    78 [ label = "feature std" ]
    79 [ label = "feature std" ]
    80 [ label = "feature proc-macro" ]
    81 [ label = "feature unicode" ]
    82 [ label = "feature unicode-age" ]
    83 [ label = "feature unicode-bool" ]
    84 [ label = "feature unicode-case" ]
    85 [ label = "feature unicode-gencat" ]
    86 [ label = "feature unicode-perl" ]
    87 [ label = "feature unicode-script" ]
    88 [ label = "feature unicode-segment" ]
    89 [ label = "feature std" ]
    90 [ label = "feature ntsecapi" ]
    91 [ label = "feature wtypesbase" ]
    92 [ label = "feature proc-macro" ]
    93 [ label = "feature spans" ]
    94 [ label = "feature default" ]
    95 [ label = "feature default" ]
    96 [ label = "feature default" ]
    97 [ label = "feature default" ]
    98 [ label = "feature default" ]
    99 [ label = "feature full" ]
    100 [ label = "feature MessageEvent" ]
    101 [ label = "feature Worker" ]
    102 [ label = "feature spans" ]
    103 [ label = "feature visit" ]
    104 [ label = "feature spans" ]
    105 [ label = "feature leftpad" ]
    106 [ label = "feature leftier-strings" ]
    107 [ label = "feature lazy_static" ]
    108 [ label = "feature libloading" ]
    109 [ label = "feature clang_5_0" ]
    110 [ label = "feature clang_4_0" ]
    111 [ label = "feature clang_3_9" ]
    112 [ label = "feature clang_3_8" ]
    113 [ label = "feature clang_3_7" ]
    114 [ label = "feature clang_3_6" ]
    115 [ label = "feature clang_3_5" ]
    116 [ label = "feature audio_toolbox" ]
    117 [ label = "feature audio_unit" ]
    118 [ label = "feature core_audio" ]
    119 [ label = "feature open_al" ]
    120 [ label = "feature core_midi" ]
    121 [ label = "feature std" ]
    122 [ label = "feature alloc" ]
    123 [ label = "feature alloc" ]
    124 [ label = "feature race" ]
    125 [ label = "feature unicode-age" ]
    126 [ label = "feature unicode-bool" ]
    127 [ label = "feature unicode-case" ]
    128 [ label = "feature unicode-gencat" ]
    129 [ label = "feature unicode-perl" ]
    130 [ label = "feature unicode-script" ]
    131 [ label = "feature unicode-segment" ]
    132 [ label = "feature alloc" ]
    133 [ label = "feature dev_urandom_fallback" ]
    134 [ label = "feature once_cell" ]
    135 [ label = "feature std" ]
    136 [ label = "feature std" ]
    137 [ label = "feature derive" ]
    138 [ label = "feature parsing" ]
    139 [ label = "feature printing" ]
    140 [ label = "feature clone-impls" ]
    141 [ label = "feature proc-macro" ]
    142 [ label = "feature quote" ]
    143 [ label = "feature std" ]
    144 [ label = "feature spans" ]
    145 [ label = "feature std" ]
    146 [ label = "feature EventTarget" ]
    147 [ label = "feature Event" ]
    0 -> 52 [ label = "" ]
    0 -> 53 [ label = "(dev) 'cfg(target_arch = \"x86_64\")'" ]
    0 -> 37 [ label = "" ]
    0 -> 43 [ label = " 'cfg(all(target_arch = \"wasm32\", target_feature = \"atomics\"))'" ]
    1 -> 54 [ label = "" ]
    1 -> 7 [ label = "" ]
    1 -> 55 [ label = "" ]
    1 -> 56 [ label = "" ]
    1 -> 16 [ label = "" ]
    1 -> 17 [ label = "" ]
    1 -> 27 [ label = "" ]
    1 -> 28 [ label = "" ]
    1 -> 29 [ label = "" ]
    1 -> 57 [ label = "" ]
    1 -> 58 [ label = "" ]
    1 -> 59 [ label = "" ]
    1 -> 60 [ label = "" ]
    4 -> 5 [ label = "(build)" ]
    4 -> 61 [ label = " 'x86_64-apple-darwin'" ]
    4 -> 62 [ label = "(dev)" ]
    4 -> 16 [ label = " 'cfg(any(target_os = \"android\", target_os = \"freebsd\", target_os = \"linux\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    4 -> 18 [ label = "" ]
    4 -> 19 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    4 -> 24 [ label = " 'x86_64-unknown-linux-gnu'" ]
    4 -> 35 [ label = " 'cfg(all(any(target_arch = \"aarch64\", target_arch = \"arm\", target_arch = \"x86\", target_arch = \"x86_64\"), not(target_os = \"ios\")))'" ]
    4 -> 37 [ label = "" ]
    4 -> 63 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    4 -> 64 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    4 -> 48 [ label = " 'cfg(target_os = \"windows\")'" ]
    7 -> 65 [ label = "" ]
    10 -> 14 [ label = "" ]
    10 -> 14 [ label = "(build)" ]
    10 -> 19 [ label = "" ]
    10 -> 20 [ label = "" ]
    11 -> 54 [ label = "" ]
    11 -> 12 [ label = "" ]
    11 -> 66 [ label = "" ]
    11 -> 67 [ label = "" ]
    11 -> 68 [ label = "" ]
    11 -> 69 [ label = "" ]
    11 -> 70 [ label = "" ]
    12 -> 71 [ label = "(build)" ]
    15 -> 72 [ label = "" ]
    20 -> 9 [ label = " 'cfg(unix)'" ]
    20 -> 73 [ label = " 'cfg(windows)'" ]
    20 -> 74 [ label = " 'cfg(windows)'" ]
    21 -> 9 [ label = "" ]
    24 -> 54 [ label = "" ]
    24 -> 6 [ label = "(build) 'cfg(target_os = \"dragonfly\")'" ]
    24 -> 8 [ label = "" ]
    24 -> 75 [ label = "" ]
    24 -> 76 [ label = "" ]
    24 -> 77 [ label = "" ]
    25 -> 22 [ label = "" ]
    25 -> 78 [ label = "" ]
    25 -> 23 [ label = "" ]
    25 -> 79 [ label = "" ]
    28 -> 38 [ label = "" ]
    29 -> 28 [ label = "" ]
    29 -> 80 [ label = "" ]
    30 -> 31 [ label = "" ]
    30 -> 81 [ label = "" ]
    30 -> 82 [ label = "" ]
    30 -> 83 [ label = "" ]
    30 -> 84 [ label = "" ]
    30 -> 85 [ label = "" ]
    30 -> 86 [ label = "" ]
    30 -> 87 [ label = "" ]
    30 -> 88 [ label = "" ]
    32 -> 6 [ label = "(build)" ]
    32 -> 19 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    32 -> 89 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    32 -> 89 [ label = " 'cfg(any(target_os = \"dragonfly\", target_os = \"freebsd\", target_os = \"illumos\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    32 -> 35 [ label = " 'cfg(any(target_arch = \"x86\", target_arch = \"x86_64\", all(any(target_arch = \"aarch64\", target_arch = \"arm\"), any(target_os = \"android\", target_os = \"fuchsia\", target_os = \"linux\"))))'" ]
    32 -> 39 [ label = "" ]
    32 -> 63 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    32 -> 64 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    32 -> 90 [ label = " 'cfg(target_os = \"windows\")'" ]
    32 -> 91 [ label = " 'cfg(target_os = \"windows\")'" ]
    36 -> 28 [ label = "" ]
    36 -> 80 [ label = "" ]
    36 -> 29 [ label = "" ]
    36 -> 92 [ label = "" ]
    36 -> 38 [ label = "" ]
    41 -> 9 [ label = "" ]
    41 -> 44 [ label = "" ]
    41 -> 93 [ label = "" ]
    42 -> 94 [ label = "" ]
    42 -> 21 [ label = "" ]
    42 -> 95 [ label = "" ]
    42 -> 96 [ label = "" ]
    42 -> 97 [ label = "" ]
    42 -> 98 [ label = "" ]
    42 -> 99 [ label = "" ]
    42 -> 46 [ label = "" ]
    43 -> 9 [ label = "" ]
    43 -> 15 [ label = "" ]
    43 -> 72 [ label = "" ]
    43 -> 100 [ label = " 'cfg(target_feature = \"atomics\")'" ]
    43 -> 101 [ label = " 'cfg(target_feature = \"atomics\")'" ]
    44 -> 97 [ label = "" ]
    44 -> 45 [ label = "" ]
    44 -> 102 [ label = "" ]
    45 -> 96 [ label = "" ]
    45 -> 97 [ label = "" ]
    45 -> 98 [ label = "" ]
    45 -> 103 [ label = "" ]
    45 -> 99 [ label = "" ]
    45 -> 42 [ label = "" ]
    45 -> 104 [ label = "" ]
    45 -> 46 [ label = "" ]
    47 -> 15 [ label = "" ]
    47 -> 72 [ label = "" ]
    49 -> 50 [ label = " 'i686-pc-windows-gnu'" ]
    49 -> 51 [ label = " 'x86_64-pc-windows-gnu'" ]
    71 -> 1 [ label = "" ]
    71 -> 56 [ label = "" ]
    54 -> 2 [ label = "" ]
    94 -> 3 [ label = "" ]
    105 -> 4 [ label = "" ]
    105 -> 18 [ label = "" ]
    106 -> 4 [ label = "" ]
    106 -> 105 [ label = "" ]
    107 -> 4 [ label = "" ]
    107 -> 16 [ label = "" ]
    52 -> 4 [ label = "" ]
    52 -> 105 [ label = "" ]
    56 -> 10 [ label = "" ]
    56 -> 108 [ label = "" ]
    108 -> 10 [ label = "" ]
    108 -> 20 [ label = "" ]
    55 -> 10 [ label = "" ]
    55 -> 109 [ label = "" ]
    109 -> 10 [ label = "" ]
    109 -> 110 [ label = "" ]
    110 -> 10 [ label = "" ]
    110 -> 111 [ label = "" ]
    111 -> 10 [ label = "" ]
    111 -> 112 [ label = "" ]
    112 -> 10 [ label = "" ]
    112 -> 113 [ label = "" ]
    113 -> 10 [ label = "" ]
    113 -> 114 [ label = "" ]
    114 -> 10 [ label = "" ]
    114 -> 115 [ label = "" ]
    115 -> 10 [ label = "" ]
    61 -> 11 [ label = "" ]
    61 -> 116 [ label = "" ]
    61 -> 117 [ label = "" ]
    61 -> 118 [ label = "" ]
    61 -> 119 [ label = "" ]
    61 -> 120 [ label = "" ]
    120 -> 11 [ label = "" ]
    120 -> 69 [ label = "" ]
    119 -> 11 [ label = "" ]
    119 -> 70 [ label = "" ]
    118 -> 11 [ label = "" ]
    118 -> 68 [ label = "" ]
    117 -> 11 [ label = "" ]
    117 -> 67 [ label = "" ]
    116 -> 11 [ label = "" ]
    116 -> 66 [ label = "" ]
    70 -> 12 [ label = "" ]
    69 -> 12 [ label = "" ]
    68 -> 12 [ label = "" ]
    67 -> 12 [ label = "" ]
    66 -> 12 [ label = "" ]
    62 -> 13 [ label = "" ]
    76 -> 19 [ label = "" ]
    75 -> 19 [ label = "" ]
    75 -> 121 [ label = "" ]
    121 -> 19 [ label = "" ]
    78 -> 22 [ label = "" ]
    79 -> 23 [ label = "" ]
    65 -> 25 [ label = "" ]
    65 -> 122 [ label = "" ]
    65 -> 78 [ label = "" ]
    65 -> 79 [ label = "" ]
    122 -> 25 [ label = "" ]
    89 -> 26 [ label = "" ]
    89 -> 123 [ label = "" ]
    123 -> 26 [ label = "" ]
    123 -> 124 [ label = "" ]
    124 -> 26 [ label = "" ]
    95 -> 26 [ label = "" ]
    95 -> 89 [ label = "" ]
    80 -> 28 [ label = "" ]
    96 -> 28 [ label = "" ]
    96 -> 80 [ label = "" ]
    92 -> 29 [ label = "" ]
    92 -> 80 [ label = "" ]
    97 -> 29 [ label = "" ]
    97 -> 92 [ label = "" ]
    58 -> 30 [ label = "" ]
    58 -> 125 [ label = "" ]
    58 -> 126 [ label = "" ]
    58 -> 127 [ label = "" ]
    58 -> 128 [ label = "" ]
    58 -> 129 [ label = "" ]
    58 -> 130 [ label = "" ]
    58 -> 131 [ label = "" ]
    58 -> 81 [ label = "" ]
    131 -> 30 [ label = "" ]
    131 -> 88 [ label = "" ]
    130 -> 30 [ label = "" ]
    130 -> 87 [ label = "" ]
    129 -> 30 [ label = "" ]
    129 -> 86 [ label = "" ]
    128 -> 30 [ label = "" ]
    128 -> 85 [ label = "" ]
    127 -> 30 [ label = "" ]
    127 -> 84 [ label = "" ]
    126 -> 30 [ label = "" ]
    126 -> 83 [ label = "" ]
    125 -> 30 [ label = "" ]
    125 -> 82 [ label = "" ]
    57 -> 30 [ label = "" ]
    88 -> 31 [ label = "" ]
    87 -> 31 [ label = "" ]
    86 -> 31 [ label = "" ]
    85 -> 31 [ label = "" ]
    84 -> 31 [ label = "" ]
    83 -> 31 [ label = "" ]
    82 -> 31 [ label = "" ]
    81 -> 31 [ label = "" ]
    81 -> 82 [ label = "" ]
    81 -> 83 [ label = "" ]
    81 -> 84 [ label = "" ]
    81 -> 85 [ label = "" ]
    81 -> 86 [ label = "" ]
    81 -> 87 [ label = "" ]
    81 -> 88 [ label = "" ]
    53 -> 32 [ label = "" ]
    53 -> 132 [ label = "" ]
    53 -> 133 [ label = "" ]
    133 -> 32 [ label = "" ]
    133 -> 134 [ label = "" ]
    134 -> 32 [ label = "" ]
    134 -> 26 [ label = "" ]
    132 -> 32 [ label = "" ]
    59 -> 33 [ label = "" ]
    59 -> 135 [ label = "" ]
    135 -> 33 [ label = "" ]
    60 -> 34 [ label = "" ]
    60 -> 136 [ label = "" ]
    136 -> 34 [ label = "" ]
    103 -> 36 [ label = "" ]
    99 -> 36 [ label = "" ]
    98 -> 36 [ label = "" ]
    98 -> 137 [ label = "" ]
    98 -> 138 [ label = "" ]
    98 -> 139 [ label = "" ]
    98 -> 140 [ label = "" ]
    98 -> 141 [ label = "" ]
    141 -> 36 [ label = "" ]
    141 -> 80 [ label = "" ]
    141 -> 92 [ label = "" ]
    140 -> 36 [ label = "" ]
    139 -> 36 [ label = "" ]
    139 -> 142 [ label = "" ]
    142 -> 36 [ label = "" ]
    142 -> 29 [ label = "" ]
    138 -> 36 [ label = "" ]
    137 -> 36 [ label = "" ]
    77 -> 40 [ label = "" ]
    77 -> 143 [ label = "" ]
    143 -> 40 [ label = "" ]
    72 -> 41 [ label = "" ]
    72 -> 144 [ label = "" ]
    72 -> 145 [ label = "" ]
    145 -> 41 [ label = "" ]
    144 -> 41 [ label = "" ]
    144 -> 93 [ label = "" ]
    104 -> 42 [ label = "" ]
    93 -> 44 [ label = "" ]
    93 -> 102 [ label = "" ]
    102 -> 45 [ label = "" ]
    102 -> 104 [ label = "" ]
    101 -> 47 [ label = "" ]
    101 -> 146 [ label = "" ]
    146 -> 47 [ label = "" ]
    64 -> 47 [ label = "" ]
    64 -> 146 [ label = "" ]
    100 -> 47 [ label = "" ]
    100 -> 147 [ label = "" ]
    147 -> 47 [ label = "" ]
    63 -> 47 [ label = "" ]
    91 -> 49 [ label = "" ]
    90 -> 49 [ label = "" ]
    74 -> 49 [ label = "" ]
    73 -> 49 [ label = "" ]
}

